src/telemetry.ts

```diff
@@ -36,4 +49,118 @@ export class NullTelemetryReporter implements ITelemetryReporter {
 
 }
 
+export const DefaultTelemetryIntervalInMilliseconds = 10000;
+
+export class BatchTelemetryReporter {
+    private _eventBuckets: {[eventName: string]: any};
+    private _timer: NodeJS.Timer;
+
+    public constructor(private _telemetryReporter: TelemetryReporter, private _cadenceInMilliseconds: number = DefaultTelemetryIntervalInMilliseconds) {
+        this.reset();
+        this.setup();
+    }
+
+    public reportEvent(name: string, data?: any): void {
+        if (!this._eventBuckets[name]) {
+            this._eventBuckets[name] = [];
+        }
+
+        this._eventBuckets[name].push(data);
+    }
+
+    public finalize(): void {
+        this.send();
+        clearInterval(this._timer);
+    }
+
+    private setup(): void {
+        this._timer = setInterval(() => this.send(), this._cadenceInMilliseconds);
+    }
+
+    private reset(): void {
+        this._eventBuckets = {};
+    }
+
+    private send(): void {
+        for (const eventName in this._eventBuckets) {
+            const bucket = this._eventBuckets[eventName];
+            let properties = BatchTelemetryReporter.transfromBucketData(bucket);
+            this._telemetryReporter.reportEvent(eventName, properties);
+        }
+
+        this.reset();
+    }
+    /**
+     * Transfrom the bucket of events data from the form:
+     * [{
+     *  p1: v1,
+     *  p2: v2
+     * },
+     * {
+     *  p1: w1,
+     *  p2: w2
+     *  p3: w3
+     * }]
+     *
+     * to
+     * {
+     *   p1: [v1,   w1],
+     *   p2: [v2,   w2],
+     *   p3: [null, w3]
+     * }
+     *
+     *
+     * The later form is easier for downstream telemetry analysis.
+     */
+    private static transfromBucketData(bucket: any[]): {[groupedPropertyValue: string]: string} {
+        const allPropertyNamesInTheBucket = BatchTelemetryReporter.collectPropertyNamesFromAllEvents(bucket);
+        let properties = {};
+
+        // Run through all the entries, create a holder for all potential property names
+        for (const key of allPropertyNamesInTheBucket) {
+            properties[`aggregated.${key}`] = [];
+        }
+
+        // Run through again, collect the values for each property name.
+        for (const entry of bucket) {
+            for (const key of allPropertyNamesInTheBucket) {
+                properties[`aggregated.${key}`].push(entry[key] || null);
```

Comment:
If entry[key] is "" or 0 entry[key] || null will return null

Reply:
Good catch. Fixed.